crates/compiler/src/lower_to_casm.rs

```diff
@@ -0,0 +1,462 @@
+//! for traversing the parsed code elements and generating the corresponding low-level instructions
+//! that can be executed by the Cairo-M virtual machine or assembler.
+//!
+//! Only the most basic features are supported for now :
+//! - Integer literals
+//! - Addition, subtraction, multiplication
+//! - Function calls
+//! - Local variable declarations and assignments
+//! - If/else statements
+//! - Return statements
+
+use crate::ast::*;
+use crate::casm::{CasmInstruction, CasmInstructionType};
+use std::collections::HashMap;
+
+/// The main compiler/lowering struct for converting AST code elements into CASM instructions.
+///
+/// The `Compiler` maintains state for local variables, stack frame offsets, and label generation.
+/// It provides methods for compiling expressions, statements, and functions, emitting CASM instructions
+/// as it traverses the AST.
+pub struct Compiler {
+    /// The list of top-level code elements (functions, statements, etc.) to compile
+    code_elements: Vec<CodeElement>,
+    /// The list of generated CASM instructions
+    casm_instructions: Vec<CasmInstruction>,
+    /// Mapping from local variable names to their frame pointer offsets
+    local_variables: HashMap<String, i32>,
+    /// The number of local variables currently allocated
+    local_variables_count: u64,
+    /// The current frame pointer offset (top of stack)
+    fp_offset: u64,
+    /// Counter for generating unique labels (for control flow)
+    label_counter: u64,
+}
+
+impl Compiler {
+    /// Create a new compiler instance from a list of code elements (AST root).
+    pub fn new(code_elements: Vec<CodeElement>) -> Self {
+        Self {
+            code_elements,
+            casm_instructions: Vec::new(),
+            local_variables: HashMap::new(),
+            local_variables_count: 0,
+            fp_offset: 0,
+            label_counter: 0,
+        }
+    }
+
+    /// Compile all code elements into a vector of CASM instructions.
+    ///
+    /// This is the main entry point for lowering a program.
+    pub fn compile(&mut self) -> Vec<CasmInstruction> {
+        for code_element in self.code_elements.clone() {
+            self.compile_code_element(code_element);
+        }
+        self.casm_instructions.clone()
+    }
+
+    /// Compile an integer literal expression.
+    ///
+    /// Pushes the literal onto the stack and returns the frame pointer offset where it is stored.
+    fn compile_int_literal(&mut self, expr: Expr) -> u64 {
+        assert!(matches!(expr.expr_type, ExprType::IntegerLiteral));
+        let instr = CasmInstruction {
+            instruction_type: CasmInstructionType::MovFpImm,
+            label: None,
+            arg0: self.fp_offset as i32,
+            arg1: expr.token.unwrap().lexeme.parse::<i32>().unwrap(),
+            arg2: 0,
+        };
+        self.casm_instructions.push(instr);
+        self.fp_offset += 1;
+        return self.fp_offset - 1;
```

Comment:
Usually, rust return values are done without the `return` keyword by omitting the semi-colon

```suggestion
        self.fp_offset - 1
```